### PR TITLE
Multi-qubit kernel for GPU

### LINKED ISSUE #22: Multi-qubit kernel for GPU
Following #21, we should port the new multi-qubit kernels in CUDA and ROCm.

### PR DESCRIPTION
**Pull request by stavros11**
Closes #22. As we discussed this is missing the proper buffer implementation. 

Currently I use a buffer that has the same size as the full state vector as a placeholder and to make sure that the rest of the code works okay. Ideally this vector should have size `(nthreads * nsubstates,)` and we need to index it properly wo that each thread uses a different set of values. Alternatively, we could try initializing a vector of size `(nsubstates,)` within the kernel but I searched a little bit and I am not sure if this is a good practice in CUDA.

### CODE REVIEW

**Comment by mlazzarin** [comment]
To sum up, we can proceed with different options:
 1. **Duplicate the state vector and use it as a buffer**: the option implemented in the PR. Simple to write and maintain, we lose one maximum qubit.
 2. **Dynamically allocate buffer inside the kernels**: overhead for the allocation, it raises memory issues when too much memory is allocated inside the kernel (unless we raise the heap size limit with CUDA)
 3. **Implement hard-coded kernels with a local buffer each**: it works only for three-qubit gates, then raises memory issues (at least in my configuration). We may keep the three-qubit implementation.
 4. **Allocate the buffer in shared memory within each block of threads**: the shared memory is fast because is on-chip, but the problem is that it is something like 40-60 KB, and we need to allocate ``nthreads_per_block * 2^ntargets * 16 B``, so we need to reduce a lot the number of threads per block.
 5. **Allocate the buffers in global memory and use it in each thread**: similar to the duplication of the state vector, we need to allocate a buffer of ``nthreads*2^ntargets``. If we use ``nthreads=nstates``, this is a problem, but we may implement the batch computation of #25 (which seems a bit slower but not much) and reduce the number of threads to something like 5000-10000 to solve the memory issues. We may implement the batch computation **only** for n-qubit gates, n > 3.

To me it seems like the option 5 + 3 (for three-qubit gates) is the best one. I'll provide some temporary benchmarks soon.

**Comment by mlazzarin** [comment]
I've performed additional benchmarks with a more powerful machine.

<details><summary>29 qubits: simulation times</summary>

ntargets | Simulation time CPU | Simulation time CPU Qiskit | Simulation time GPU 1 | Simulation time GPU 2 | Simulation time GPU 3 | Simulation time GPU 4 | Simulation time GPU 5 | Simulation time GPU Qiskit
-- | -- | -- | -- | -- | -- | -- | -- | --
3 | 4.37162 | 4.41568 | 0.87840 | 12.97101 | 0.83416 | 0.85116 | 0.87027 | 5.16430
4 | 4.48415 | 4.54203 | 0.98535 | 7.02323 | 0.92451 | 0.96455 | 1.00448 | 5.16169
5 | 5.48248 | 4.81841 | 1.19414 | 4.33198 | 1.10545 | 1.54272 | 1.27370 | 5.15485
6 | 5.86040 | 5.36233 | 1.63869 | 6.02275 | nan | 2.28142 | 3.12753 | 5.14882
7 | 6.09441 | 6.27977 | 2.52201 | 10.72576 | nan | nan | 5.70711 | 5.71034
8 | 7.32556 | 10.95151 | 4.29162 | nan | nan | nan | 11.78047 | 7.19121
9 | 10.10311 | 18.89987 | 8.24116 | nan | nan | nan | 24.46231 | 10.23238
10 | 15.76016 | 65.62365 | 17.28352 | nan | nan | nan | 47.32037 | 16.72979
11 | 31.98329 | 430.08204 | 37.82665 | nan | nan | nan | 94.30802 | not performed
12 | 110.54395 | 1236.13424 | 72.91512 | nan | nan | nan | 192.27257 | not performed


</details>

<details><summary>30 qubits: simulation times</summary>

ntargets | Simulation time CPU | Simulation time CPU Qiskit | Simulation time GPU 1 | Simulation time GPU 2 | Simulation time GPU 3 | Simulation time GPU 4 | Simulation time GPU 5 | Simulation time GPU Qiskit
-- | -- | -- | -- | -- | -- | -- | -- | --
3 | 8.98304 | 9.15258 | 1.79254 | 26.01708 | 1.71656 | 1.73892 | 1.78816 | 10.27811
4 | 9.24200 | 9.34022 | 2.01931 | 14.17244 | 1.89794 | 1.92429 | 2.05220 | 10.29524
5 | 11.26570 | 9.90266 | 2.43439 | 8.67897 | 2.26075 | 2.58267 | 2.59685 | 10.31337
6 | 12.43468 | 11.00892 | 3.31824 | 11.46821 | nan | 4.52018 | 6.30734 | 10.32503
7 | 12.49654 | 12.84493 | 5.09171 | 21.56444 | nan | nan | 11.31291 | 11.50012
8 | 14.95679 | 22.73134 | 8.53706 | nan | nan | nan | 23.57340 | 14.40164
9 | 20.52620 | 39.12058 | 16.06461 | nan | nan | nan | 48.98731 | 20.48101
10 | 31.42744 | 130.09857 | 32.26213 | nan | nan | nan | 94.83861 | 32.93433
11 | 63.51460 | 832.37259 | 67.39161 | nan | nan | nan | 188.75589 | not performed
12 | 201.67062 | 2415.66124 | 147.86115 | nan | nan | nan | 384.64494 | not performed

</details>

<details><summary>18 qubits: dry run time - simulation time</summary>

ntargets | delta CPU | delta CPU Qiskit | delta GPU 1 | delta GPU 2 | delta GPU 3 | delta GPU 4 | delta GPU 5 | delta GPU Qiskit
-- | -- | -- | -- | -- | -- | -- | -- | --
3 | 0.13864 | 0.02050 | 0.66376 | 0.55524 | 0.85127 | 0.78859 | 0.84690 | 0.00309
4 | 0.16576 | 0.02005 | 0.66619 | 0.55557 | 0.85902 | 0.78326 | 0.86043 | 0.00299
5 | 0.13019 | 0.01383 | 0.66636 | 0.55317 | 0.86737 | 0.77655 | 0.86716 | 0.00244
6 | 0.14925 | 0.12211 | 0.65860 | 0.55048 | nan | 0.66763 | 0.85594 | 0.00265
7 | 0.13032 | 0.13500 | 0.66649 | 0.54820 | nan | nan | 0.84973 | 0.00364
8 | 0.13304 | 0.11849 | 0.66336 | 0.55268 | nan | nan | 0.85293 | 0.01142
9 | 0.14281 | 0.07636 | 0.68302 | 0.55383 | nan | nan | 0.86233 | 0.05565
10 | 0.14997 | 0.07241 | 0.65187 | 0.55876 | nan | nan | 0.85332 | 0.03711
11 | 0.13002 | 0.04130 | 0.70330 | 0.56513 | nan | nan | 0.88719 | not performed
12 | 0.14162 | -0.06923 | 0.74065 | 0.59264 | nan | nan | 0.92795 | not performed

</details>

I've managed to use hard-coded kernels up to 5 targets (option 3) by reducing the number of threads per block (just for those specific calls). It is the fastest option, even though the overhead of the dry run increases a bit.

Note the Qiskit seems very slow with more than 10 targets. Also the results with the GPU are a bit strange. Maybe there's something wrong with my implementation, I don't know. I had to skip the results for the GPU with n>10 due to time limits.

Also the results with option 2 are a bit weird, I'll double-check them tomorrow.

Given these results, I'd say that the best option is to use hard-coded kernels for n<=5 (option 3) and duplicate the state vector for n>5 (option 1). Moreover, in this way we don't need to specify an hard-coded number of blocks (while we still need to hard-code the number of threads).

File: src/qibojit/custom_operators/backends.py [hunk 1]
```diff
@@ -268,6 +284,47 @@ def two_qubit_base(self, state, nqubits, target1, target2, kernel, qubits=None,
...
+        # The latter require a full copy ``buffer``` of the state vector
+        ktype = self.get_kernel_type(state)
+        if len(targets) < 6:
+            # Compute the number of blocks and threads
+            # To avoid memory issues, reduce the block size with len(targets)
+            # len(targets): 3 -> 1024 threads per block
+            # len(targets): 4 -> 512  threads per block
+            # len(targets): 5 -> 256  threads per block
+            nblocks, block_size = self.calculate_blocks(nstates,
+                                                        block_size=2**(13-len(targets)))
```
[the remark this answers is not in the document]
Comment (mlazzarin) [reply src/qibojit/custom_operators/backends.py:309]: Ok, I confirm that it is not independent of GPU model, as this configuration doesn't work with a NVIDIA Tesla V100 (I need to halve the block size). I need to find a smarter way to manage this before merging this PR.

[the remark this answers is not in the document]
Comment (mlazzarin) [reply src/qibojit/custom_operators/backends.py:309]: I think that I've found a smarter way in #37 .

**Comment by scarrazza** [comment]
@mlazzarin @stavros11 could you please fix the conflicts so we can merge? Thanks.

**Comment by mlazzarin** [comment]
@stavros11 thank you for fixing the conflicts.

**Comment by stavros11** [comment]
@mlazzarin @scarrazza before merging this, please make sure that tests pass in your configurations with the GPUs enabled and disabled, particularly for AMDs since I do not have access to these and the CI is not testing for GPU. This looks good to me, my only concerns are the following:
* The multi-qubit kernels increase the dry run/compilation time by around 0.3sec which is small but non-negligible and happens regardless of whether the user actually uses these kernels or no.
* The tests for 22 qubits take some time, at least in my machine, but since they don't break the CI it is not a big issue.

Just mentioning these to keep in mind, I am not sure if we should/can do something in the current stage.

**Comment by mlazzarin** [comment]
> @mlazzarin @scarrazza before merging this, please make sure that tests pass in your configurations with the GPUs enabled and disabled, particularly for AMDs since I do not have access to these and the CI is not testing for GPU.

Ok, I'm on it.

> The multi-qubit kernels increase the dry run/compilation time by around 0.3sec which is small but non-negligible and happens regardless of whether the user actually uses these kernels or no.

We may solve this issue by redesigning the way we load the kernels, e.g. by replacing ``cp.RawModule`` with many ``cp.RawKernel`` and compile each kernel only when it's needed. 

> The tests for 22 qubits take some time, at least in my machine, but since they don't break the CI it is not a big issue.

The tests for 22 qubits are required to catch some resource issues with GPUs. Given that the CI doesn't use a GPU, is it possible the implement some skipping mechanism that works only if a GPU is not available?

**Comment by mlazzarin** [comment]
Ok, I tried with both Intel/AMD CPUs and NVIDIA/AMD GPUs and the qibojit tests are fine.

**Comment by scarrazza** [comment]
@stavros11 thanks for the summary, @mlazzarin thanks for testing. Let me suggest to merge and release a 0.0.4rc0, then next week we can discuss about this compilation issue with their developers.

**Pull request merged**